File: code/js/RecursiveHoFs.js

var R = {
	
	each: function(list, op) {
		if(list.length > 0) {
			op(list[0]);
			R.each(list.slice(1), op);
		}
	},
	
	map: function(list, op) { return R._map(list, [], op); },
	
	_map: function(list, acc, op) {
		if(list.length == 0)
			return acc;
		else {
			acc.push(op(list[0]));
			return R.map(list.slice(1), acc, op);
		}
	},
	
	filter: function(list, pred) { return R._filter(list, [], pred); },
	
	_filter: function(list, acc, pred) {
		if(list.length == 0) 
			return acc;
		else {
			if(pred(list[0]))
				acc.push(list[0]);
			return R._filter(list.slice(1), acc, pred);
		}
	},
	
	foldr: function(list, initial, op) {
		if(list.length == 0)
			return initial;
		else
			return R.foldr(list.slice(1), op(initial, list[0]), op);
	}
	
}

File: code/scala/RecursiveHoFs.scala
/*
 * Things are a bit better in scala, because we can use pattern matching for
 * painless list extractions and a nice shorthand for anonymous functions.
 *
 * Chris Lewis (chris@thegodcode.net), for JaxFunc (http://www.meetup.com/JaxFunc/)
 */
object RF {
  
  def foldr[A, B](f: (B, A) => B, init: B, list: List[A]): B = list match {
    case Nil => init
    case head :: tail => foldr(f, f(init, head), tail)
  }
  
  def each[A](list: List[A], op: A => Unit) = 
    foldr((init: List[A], a: A) => { op(a); init }, Nil, list)
  
  def map[A, B](list: List[A], f: A => B) =
    foldr((l: List[B], a: A) => f(a) :: l, Nil, list)
  
  def filter[A](list: List[A], pred: A => Boolean) = 
    foldr((l: List[A], a: A) => if(pred(a)) a :: l else l, Nil, list)
  
}
